portfolio/src/main/java/com/google/sps/data/CommentAnalysis.java

```diff
@@ -0,0 +1,136 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.BatchAnnotateImagesResponse;
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.cloud.vision.v1.Feature;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.gson.Gson;
+import com.google.protobuf.ByteString;
+
+/**
+ * Represents analysis with GCloud of a particular comment. Right now just holds image labels, but
+ * will eventually also analyze the sentiment of the comment text
+ */
+public class CommentAnalysis {
+  private List<ImageLabel> imageLabels;
+
+  /** Represents a single label for an image. Serves as a convenient wrapper for serialization */
+  private static class ImageLabel {
+    private final String description;
+    private final float score;
+
+    public ImageLabel(EntityAnnotation annotation) {
+      description = annotation.getDescription();
+      score = annotation.getScore();
+    }
+
+    public String getDescription() {
+      return description;
+    }
+
+    public float getScore() {
+      return score;
+    }
+  }
+
+  public CommentAnalysis() {
+    imageLabels = new ArrayList<>();
+  }
+
+  /**
+   * Attaches image analysis, reading from the given comment. Performs no analysis if image does not
+   * exist
+   */
+  public void analyzeImage(Comment comment) throws IOException {
+    if (comment.getBlobKey() == null) {
+      return;
+    }
+    byte[] imageBytes = getBlobBytes(comment.getBlobKey());
+    List<EntityAnnotation> labels = getImageLabels(imageBytes);
```

Comment:
Always consider immutable list first unless have a good reason not to.
go/java-practices/choosing-collections#immutable

Reply:
Makes sense -- `getImageLabels` returns directly from a GCloud API, so should I wrap that with an immutable list? I can't access the link you sent me to confirm :(